[iOS] Add support for keyboard shortcut and menu item to find using the current selection

pxlcoder opened the pull request:
#### e921f41ad5bcd9d1f27b55d1a13c95b586113b29
<pre>
[iOS] Add support for keyboard shortcut and menu item to find using the current selection
<a href="https://bugs.webkit.org/show_bug.cgi?id=242196">https://bugs.webkit.org/show_bug.cgi?id=242196</a>
rdar://93530469

Reviewed by Wenson Hsieh.

This patch adds support for:
- The CMD+E keyboard shortcut, which populates the search text for the current find interaction
- A &quot;Find Selection&quot; item in the callout bar, which populates the search text and presents the find panel

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView canPerformActionForWebView:withSender:]):
(-[WKContentView useSelectionForFindForWebView:]):
(-[WKContentView _findSelectedForWebView:]):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::setTextIndicatorWithRange):

Fix the appearance of the highlighted range in cases where it is equal to the
current selection.

If the selection before enabling appearance updates is equal to the range to
highlight, the TemporarySelectionChange used by TextIndicator to generate a
snapshot will exit early and fail to call updateAppearance. This means that
the selection won&apos;t be pushed to the render tree, and the TextIndicator will
have a blank snapshot.

This logic matches the existing update logic in FindControllerIOS.

Canonical link: <a href="https://commits.webkit.org/252144@main">https://commits.webkit.org/252144@main</a>
</pre>

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -4107,6 +4107,16 @@ - (BOOL)canPerformActionForWebView:(SEL)action withSender:(id)sender
+        if (!self.webView._findInteractionEnabled)
+            return NO;
+
+        if (!self.selectedText.length)
```
Nit - `selectedText` is a bit of a misnomer here, since it can return a non-empty string even in the case of a caret selection in editable content (if the caret is next to a word).

Perhaps it would be more appropriate to check `_page->editorState().selectionIsRange` as well?

pxlcoder replied:
Yes! Updated.

pxlcoder committed:
[iOS] Add support for keyboard shortcut and menu item to find using the current selection
https://bugs.webkit.org/show_bug.cgi?id=242196
rdar://93530469

Reviewed by Wenson Hsieh.

This patch adds support for:
- The CMD+E keyboard shortcut, which populates the search text for the current find interaction
- A "Find Selection" item in the callout bar, which populates the search text and presents the find panel

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView canPerformActionForWebView:withSender:]):
(-[WKContentView useSelectionForFindForWebView:]):
(-[WKContentView _findSelectedForWebView:]):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::setTextIndicatorWithRange):

Fix the appearance of the highlighted range in cases where it is equal to the
current selection.

If the selection before enabling appearance updates is equal to the range to
highlight, the TemporarySelectionChange used by TextIndicator to generate a
snapshot will exit early and fail to call updateAppearance. This means that
the selection won't be pushed to the render tree, and the TextIndicator will
have a blank snapshot.

This logic matches the existing update logic in FindControllerIOS.

Canonical link: https://commits.webkit.org/252144@main
